mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,745 @@
+    #[snafu(display("Incorrect MetaDataIndex checksum"))]
+    IndexChecksum {},
+}
+
+#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct MetaDataChildEntry {
+    pub guid: Guid,
+    pub state: u16,
+}
+
+impl MetaDataChildEntry {
+    /// Calculate checksum over the child entries
+    pub fn checksum(children: &[MetaDataChildEntry]) -> u32 {
+        let mut digest = crc32::Digest::new(crc32::IEEE);
+        let count = children.len() as u32;
+        for child in children {
+            digest.write(&serialize(child).unwrap());
+        }
+        if count < MetaDataObject::MAX_CHILD_ENTRIES {
+            let pad = serialize(&MetaDataChildEntry::default()).unwrap();
+            for _ in count .. MetaDataObject::MAX_CHILD_ENTRIES {
+                digest.write(&pad);
+            }
+        }
+        digest.sum32()
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+struct MetaDataHeader {
+    signature: [u8; 8],
+    self_checksum: u32,
+    generation: u64,
+    timestamp: u128,
+    table_crc: u32,
+    count: u32,
+}
+
+impl MetaDataHeader {
+    /// Checksum the header with the checksum field itself set to 0
+    fn checksum(&mut self) -> u32 {
+        self.self_checksum = 0;
+        self.self_checksum = crc32::checksum_ieee(&serialize(&self).unwrap());
+        self.self_checksum
+    }
 }
 
-#[derive(Debug, Deserialize, PartialEq, Default, Serialize, Copy, Clone)]
-pub struct MetaDataHeader {
-    /// Signature identifying this as a MetaDataHeader object
+#[derive(Debug, PartialEq)]
+pub struct MetaDataObject {
     pub signature: [u8; 8],
-    /// Size of this header in bytes
-    pub header_size: u32,
-    /// CRC-32 checksum of this header
     pub self_checksum: u32,
-    /// Current object generation counter
     pub generation: u64,
-    /// Absolute location (LBA) of this header on disk
-    pub self_lba: u64,
-    /// Offset of start of index table relative to self_lba
-    pub index_start: u64,
-    /// Number of valid entries in the index
-    pub used_entries: u32,
-    /// Maximum number of entries that the index can contain
-    pub max_entries: u32,
-    /// Size of an index entry in bytes
-    pub entry_size: u32,
-    /// CRC-32 checksum of the index table
-    pub index_checksum: u32,
-    /// Offset of first usable data block relative to self_lba
-    pub data_start: u64,
-    /// Offset of last usable data block relative to self_lba
-    pub data_end: u64,
+    pub timestamp: u128,
+    pub table_crc: u32,
+    pub children: Vec<MetaDataChildEntry>,
 }
 
-impl MetaDataHeader {
-    pub const METADATA_HEADER_SIZE: u32 = 72;
-    pub const MAX_INDEX_ENTRIES: u32 = 32;
-    pub const INDEX_ENTRY_SIZE: u32 = 44;
+impl MetaDataObject {
+    /// Observe that sizeof(MetaDataHeader) + 26 * sizeof(MetaDataChildEntry) is
+    /// exactly 512. While it may be smaller, the value of this expression
+    /// must NEVER exceed 512.
+    pub const MAX_CHILD_ENTRIES: u32 = 26;
```

Comment:
Would be good to have common definition of this constant and use it also in add-nexus-child method in order to be consistent. I assume that nothing prevents grpc client to add more than 26 children at this point. But maybe we want to enforce much lower limit at the grpc level. Not sure. In any case the count should never go beyond 26.

Reply:
When we create the nexus, we should add the first child states, so we should get a validation error which could then fail the creation altogether?

Reply:
From the consistency perspective, it would be nice for this constant to be explicitly expressed via sizes of all the participating structures: that would allow facing errors when we add some new fields to them. Also, would be nice to have some defensive logic (assertions, etc) to guarantee that the total size never goes beyond 512 (if such defensive checks already exist in the code, sorry for overlooking).

Reply:
There is a checks insofar as any attempt to add a MetaDataObject that had more than 26 child entries will fail. However as far as I am aware, there are no other checks elsewhere to enforce any limit on the number of children that a Nexus may have.